client/buffered_client.go

```diff
@@ -0,0 +1,131 @@
+// FlushMaxPoints: Buffer at most this many points in memory before flushing to the server.
+// FlushMaxWaitTime: Buffer points in memory for at most this amount of time before flushing to the server.
+type BufferConfig struct {
+	Database         string
+	FlushMaxPoints   int
+	FlushMaxWaitTime time.Duration
+}
+
+// NewBufferedClient will instantiate and return a connected BufferedClient.
+func NewBufferedClient(clientConfig Config, bufferConfig BufferConfig) (bufferedClient *BufferedClient, err error) {
+	client, err := NewClient(clientConfig)
+	if err != nil {
+		return
+	}
+	bufferedClient = &BufferedClient{
+		Client:       client,
+		bufferConfig: bufferConfig,
+		ingestChan:   make(chan Point, bufferConfig.FlushMaxPoints/3),
+		closeChan:    make(chan chan error, 1),
+		flushTimer:   time.NewTimer(bufferConfig.FlushMaxWaitTime),
+		pointsBuf:    make([]Point, bufferConfig.FlushMaxPoints),
+		pointsIndex:  0,
+	}
+	go bufferedClient.ingestAndFlushLoop()
+	return
+}
+
+// BufferedClient is used to buffer points in memory and periodically flush them to the server
+type BufferedClient struct {
+	*Client
+	bufferConfig BufferConfig
+	ingestChan   chan Point
+	closeChan    chan chan error
+	flushTimer   *time.Timer
+	pointsBuf    []Point
+	pointsIndex  int
+}
+
+// Add a Point with the given values to the BufferedClient.
+// If the BufferedClient is closed, didAdd is false
+func (b *BufferedClient) Add(measurement string, val interface{}, tags map[string]string, fields map[string]interface{}) (didAdd bool) {
+	ingestChan := b.ingestChan
+	if ingestChan == nil {
+		return
+	}
+	if fields == nil {
+		fields = make(map[string]interface{}, 1)
+	}
+	fields["value"] = val
+	ingestChan <- Point{
+		Measurement: measurement,
+		Tags:        tags,
+		Fields:      fields,
+		Time:        time.Now(),
+	}
+	didAdd = true
+	return
+}
+
+// Close will close the BufferedClient. While closing, it will flush any points from Add()
+// This method executes asynchronously, but it returns a channel which can be read from to ensure that the buffered client
+// Once the client
+func (b *BufferedClient) Close() error {
+	closeResultChan := make(chan error)
+	b.closeChan <- closeResultChan
+	return <-closeResultChan
+}
+
+// Async ingest and flush loop
+//////////////////////////////
+
+// Read ingested points, buffer them in memory, and periodically flush to server.
+// On Close(), drain ingested points, flush to server, and signal that Close has completed.
+func (b *BufferedClient) ingestAndFlushLoop() {
+	for b.ingestChan != nil {
+		select {
+		case point := <-b.ingestChan:
+			b.processIngestedPoint(point)
+		case <-b.flushTimer.C:
+			b.flushBatch()
+		case closeResultChan := <-b.closeChan:
+			ingestChan := b.ingestChan
+			b.ingestChan = nil // At this point b.Add() becomes a no-op and starts returning false
+			b.drainChan(ingestChan)
+			b.flushBatch()
+			closeResultChan <- nil
```

Comment:
I'm not sure but I think you need to stop your timer in here as well or that go routine will continue running.